### PR TITLE
vault secrets file, keep context when symlink

### LINKED ISSUE #18319: Symlinked executable vault password files should be invoked by their original path, not real path
##### ISSUE TYPE
 - Bug Report

##### COMPONENT NAME
`--vault-password-file`

##### ANSIBLE VERSION
```
ansible 1.9.3
```
(but also occurs in Ansible 2)

##### DESCRIPTION
When specifying an executable vault password file, Ansible invokes it by its "real" path, not by the path as specified. This matters if what's passed is actually a *symlink* to an executable, and the executable uses the path by which it is invoked to determine the password to produce.

For example:

```
$ ls -l
total 16
-rwxr-xr-x  1 julian  staff  290 Nov  2 23:03 get-password
-rw-------  1 julian  staff  355 Nov  2 23:10 secrets.yml
```

`secrets.yml` is a dummy variables vault file created by `ansible-vault create` with the password `foo-bar-ansible`.

```
$ cat get-password
#!/bin/sh

basename="$(basename $0)"
basename_prefix="get-password"
default_password="foo-bar"

case "${basename}" in
  "${basename_prefix}"-*)
    password="${default_password}-${basename#${basename_prefix}-}"
    ;;
  *)
    password="${default_password}"
    ;;
esac

echo "${password}"
```

This script produces the password `foo-bar` if invoked directly, but produces a different password when invoked through a symlink that begins with `get-password-`, e.g., `foo-bar-quux` when invoked as `get-password-quux`. This script is a synthetic example, but I actually use this mechanism with a more complicated password retrieval script that installs various symlinks for different applications, Ansible being one of them.

Now lets create a symlink with a suitable name:

```
$ ln -s get-password get-password-ansible
$ ls -l get-password*
-rwxr-xr-x  1 julian  staff  290 Nov  2 23:03 get-password
lrwxr-xr-x  1 julian  staff   12 Nov  2 23:24 get-password-ansible -> get-password
$ ./get-password
foo-bar
$ ./get-password-ansible
foo-bar-ansible
$ ansible all -i localhost, -c local -a uptime -e @secrets.yml --vault-password-file get-password-ansible
ERROR: Decryption failed
```

Oops. We're invoking `get-password-ansible`, which as we just saw, produces the correct password, `foo-bar-ansible`. What gives?

Well, now let's try replacing the symlink with a straight-up copy of the real file:

```
$ rm -f get-password-ansible
$ cp get-password get-password-ansible
$ ls -l get-password*
-rwxr-xr-x  1 julian  staff  290 Nov  2 23:03 get-password
-rwxr-xr-x  1 julian  staff  290 Nov  2 23:12 get-password-ansible
$ ./get-password
foo-bar
$ ./get-password-ansible
foo-bar-ansible
$ ansible all -i localhost, -c local -a uptime -e @secrets.yml --vault-password-file get-password-ansible
localhost | success | rc=0 >>
23:12  up 29 days,  2:46, 24 users, load averages: 1.22 1.50 1.58
```

Oh, hey, this time it worked!

And this is consistent with how the `real_vault_password_file` function is defined in https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/__init__.py#L613. It calls `os.path.realpath()` on the file name provided and then invokes the script by that name.

It would be more useful if the vault password file's executability was determined by its real path, etc., but then invoked by its *original* path.

### PR DESCRIPTION
**Pull request by bcoca**
fixes #18319

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

vault

### CODE REVIEW

File: lib/ansible/parsing/vault/__init__.py [hunk 1]
```diff
@@ -349,17 +349,19 @@ def script_is_client(filename):
 
 
 def get_file_vault_secret(filename=None, vault_id=None, encoding=None, loader=None):
-    this_path = os.path.realpath(os.path.expanduser(filename))
 
+    # while files from --vault-password-file are unfracked, other sources are not
+    this_path = unfrackpath(filename, follow=False)
```
Comment (s-hertel) [inline lib/ansible/parsing/vault/__init__.py:358]: It's already been through unfrackpath once at this point in our code, but I guess this would be helpful for other callers.

File: changelogs/fragments/vault_syml_allow.yml [hunk 2]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - vault secrets file now executes in the correct context when it is a symlink (not resolved to original file).
```
Comment (s-hertel) [inline changelogs/fragments/vault_syml_allow.yml]: In the issue 'original path' was used to describe the symlink path rather than the realpath.
```suggestion
  - vault secrets file now executes in the correct context when it is a symlink (not resolved to canonical file).
```

**Review by s-hertel — APPROVED** [review]

File: lib/ansible/parsing/vault/__init__.py (hunk 1 above)
Comment (bcoca) [reply lib/ansible/parsing/vault/__init__.py:358]: only when it comes from --cli, other paths bypass that (learnt the hard way)

File: changelogs/fragments/vault_syml_allow.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+bugfixes:
+  - vault secrets file now executes in the correct context when it is a symlink (not resolved to canonical file).
```
Comment (webknjaz) [inline changelogs/fragments/vault_syml_allow.yml:2]: Is that trailing period necessary when the sentence is not capitalized?

File: lib/ansible/cli/vault.py [hunk 4]
```diff
@@ -408,8 +408,7 @@ def _format_output_vault_strings(self, b_plaintext_list, vault_id=None):
             # (the text itself, which input it came from, its name)
             b_plaintext, src, name = b_plaintext_info
 
-            b_ciphertext = self.editor.encrypt_bytes(b_plaintext, self.encrypt_secret,
-                                                     vault_id=vault_id)
+            b_ciphertext = self.editor.encrypt_bytes(b_plaintext, self.encrypt_secret, vault_id=vault_id)
```
Comment (webknjaz) [inline lib/ansible/cli/vault.py:411]: This is just a formatting change, right?

File: lib/ansible/parsing/vault/__init__.py [hunk 5]
```diff
@@ -349,17 +349,19 @@ def script_is_client(filename):
 
 
 def get_file_vault_secret(filename=None, vault_id=None, encoding=None, loader=None):
-    this_path = os.path.realpath(os.path.expanduser(filename))
 
+    # while files from --vault-password-file are unfracked, other sources are not
```
Comment (webknjaz) [inline lib/ansible/parsing/vault/__init__.py]: Could you extend this comment to explain more of the motivation and not only the effect?

File: lib/ansible/parsing/vault/__init__.py [hunk 6]
```diff
@@ -659,8 +660,7 @@ def decrypt_and_get_vault_id(self, vaulttext, filename=None, obj=None):
                 msg += "%s is not a vault encrypted file" % to_native(filename)
             raise AnsibleError(msg)
 
-        b_vaulttext, dummy, cipher_name, vault_id = parse_vaulttext_envelope(b_vaulttext,
-                                                                             filename=filename)
+        b_vaulttext, dummy, cipher_name, vault_id = parse_vaulttext_envelope(b_vaulttext, filename=filename)
```
Comment (webknjaz) [inline lib/ansible/parsing/vault/__init__.py:669]: Just formatting?

File: test/integration/targets/vault/script/vault-secret.sh [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+basename="$(basename $0)"
+# shellcheck disable=SC2046
+# shellcheck disable=SC2086
+dirname="$(basename $(dirname $0))"
+basename_prefix="get-password"
+default_password="foo-bar"
+
+case "${basename}" in
+  "${basename_prefix}"-*)
+    password="${default_password}-${basename#${basename_prefix}-}"
```
Comment (webknjaz) [inline test/integration/targets/vault/script/vault-secret.sh]: It is rather hard to parse what's going on here. Have you considered using Python and not shell for this script? I bet it'd be more readable.

File: test/integration/targets/vault/aliases [hunk 8]
```diff
@@ -0,0 +1,2 @@
+shippable/posix/group4
```
Comment (s-hertel) [inline test/integration/targets/vault/aliases]: Oh, one thing I just noticed - there is already an ansible-vault test target.

File: test/integration/targets/vault/symlink/get-password-symlink [hunk 9]
```diff
@@ -0,0 +1 @@
+../script/vault-secret.sh
```
Comment (webknjaz) [inline test/integration/targets/vault/symlink/get-password-symlink]: Git tends to act funny with symlinks in some corner cases. Would you consider creating this in `runme.sh` dynamically + gitignoring it?

File: lib/ansible/parsing/vault/__init__.py (hunk 6 above)
Comment (bcoca) [reply lib/ansible/parsing/vault/__init__.py:669]: i might have done some while trying to trace where things were

File: changelogs/fragments/vault_syml_allow.yml (hunk 3 above)
Comment (bcoca) [reply changelogs/fragments/vault_syml_allow.yml:2]: idk

File: test/integration/targets/vault/symlink/get-password-symlink (hunk 9 above)
Comment (bcoca) [reply test/integration/targets/vault/symlink/get-password-symlink]: we use symlinks in other tests, what is the exact issue?

Comment (webknjaz) [reply test/integration/targets/vault/symlink/get-password-symlink]: On some systems Git turns symlinks into regular files. We won't see much of this here but it wouldn't hurt to be cautious.

**Comment by webknjaz** [comment]
Some tmpdir removals are slow in CI which caused failures. I'm going to restart those jobs since there doesn't seem to be any change related to that.

File: test/integration/targets/vault/script/vault-secret.sh (hunk 7 above)
Comment (bcoca) [reply test/integration/targets/vault/script/vault-secret.sh]: used the user supplied example, its just using the basename of the path and it's directory to generate the secret

**Comment by webknjaz** [comment]
> Some tmpdir removals are slow in CI which caused failures. I'm going to restart those jobs since there doesn't seem to be any change related to that.

Restart didn't help :(

File: test/integration/targets/ansible-vault/runme.sh [hunk 10]
```diff
@@ -549,3 +549,27 @@ grep out.txt -e "[WARNING]: Error in vault password file loading (id2)"
...
+
+# using symlink
+ansible-playbook symlink.yml "$@" --vault-password-file symlink/get-password-symlink
+
+### NEGATIVE TESTS
+
+ER='Attempting to decrypt'
+#### no secrets
+# 'real script'
+ansible-playbook realpath.yml "$@" 2>&1 |grep "${ER}"
```
Comment (webknjaz) [inline test/integration/targets/ansible-vault/runme.sh:565]: Looks like this is what's failing.

File: test/integration/targets/ansible-vault/runme.sh [hunk 11]
```diff
@@ -549,3 +549,31 @@ grep out.txt -e "[WARNING]: Error in vault password file loading (id2)"
 grep out.txt -e "ERROR! Did not find a match for --encrypt-vault-id=id2 in the known vault-ids ['id3']"
 
 set -e
+
```
Comment (s-hertel) [inline test/integration/targets/ansible-vault/runme.sh:553]:
```suggestion
unset ANSIBLE_VAULT_IDENTITY_LIST

```
